### DISCUSSION TITLE
defaultValues for fieldArray when using a formContext

### DISCUSSION THREAD

**Discussion by ranran2121** [react-hook-form/react-hook-form#9602] [Q&A]
Hi,

I have a long-form that I have decided to split it into smaller components and so I am using the Form Provider. 
In one of the components I have some inputs that I need to duplicate and so I am using the FieldArray. 
I need to have a first row with empty inputs but I have no clear where I can set the DefaultValues, either in the parent component holding the FormProvider or in the child component where I use the useFieldArray hook.
Here a sample of the code
Profile.js
```
import {
  useForm,
  FormProvider,
  useFormContext,
  useFieldArray,
} from "react-hook-form";

const MyProfile = (props) => {
 
  const methods = useForm();
 
  const onSubmit = async (formData) => {
    console.log(formData);   
  };

  return (
          <FormProvider {...methods}>
            <form onSubmit={methods.handleSubmit(onSubmit)}>    
             <Skills />
                <button type="submit" className="btn-darkgreen w-full">
                  save
                </button>             
            </form>
          </FormProvider>
  );
};
```

Skills.js
```
import React, { useEffect, useState } from "react";
import {
  useForm,
  FormProvider,
  useFormContext,
  useFieldArray,
} from "react-hook-form";

const SkillsForm = (props) => {  
  const { register, formState } = useFormContext();
  const { errors } = formState;
  const { fields, append, prepend, remove, swap, move, insert } = useFieldArray(
    {
      name: "test", 
    }
  );

 
  return (
    <>
      <div>
        <p >
          Tell us what you skills are
        </p>
        {fields.map((field, index) => {

          return (
            <div
              key={field.id} >
              <div >
                <div >
                  <label htmlFor="skills" >
                    Skill
                  </label>
                  <select
                    //defaultValue={userInfo?.residenceCountryIsoCode}
                    {...register(`test.${index}.skill`)}
                    id="skills"
                  >
                    <option value=""> Select </option>
                    {skills &&
                      skills.map((skill) => {
                        return (
                          <option value={skill.id} key={skill.name}>
                            {skill.name}
                          </option>
                        );
                      })}
                  </select>                 
                </div>
               
                  <div >
                    <label
                      htmlFor="yearsExperience"
                    >
                      Years
                    </label>
                    <input
                      type="number"
                      min="0"
                      {...register(`test.${index}.yearsExperience`)}
                      id="yearsExperience"
                    />                    
                  </div>

                  <div >                  
                    <label htmlFor="expertise" >
                      Expertise
                    </label>
                    <select                    
                      {...register(`test.${index}.expertise`)}
                      id="expertise"
                    >
                      <option value=""> Select </option>
                      {["BASIC", "INTERMEDIATE", "ADVANCED"].map((level) => {
                        return (
                          <option value={level} key={level}>
                            {level}
                          </option>
                        );
                      })}
                    </select>                  
                  </div>
               
              </div>

              <button
                onClick={() => remove(index)}               
                <XCircleIcon  />
              </button>
            </div>
          );
        })}

        <button
          onClick={() =>
            append({ skill: "", yearsExperience: 0, expertise: "" })
          }         
        >
          <PlusCircleIcon  />
          <span >add row skill</span>
        </button>
      </div>
    </>
  );
};

export default SkillsForm;

```
Can someone indicate me how to set the defaultValues so that I can display at least one row with empty fields? 
Thank you

**Comment by leapful** [comment]
@ranran2121 Could you please provide a codesandbox url to re-produce the issue ?

**Reply by ranran2121** [reply]
@leapful 
here you are. (please navigate to the /profile page)
https://codesandbox.io/p/github/codesandbox/codesandbox-template-next.js/csb-h3qilg/draft/hardcore-feather?file=%2Fpages%2Findex.js&selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A8%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A8%7D%5D&workspace=%257B%2522activeFileId%2522%253A%2522cl5s5elnx000hlohv6m80h2eb%2522%252C%2522openFiles%2522%253A%255B%2522%252Fpages%252Fprofile.js%2522%252C%2522%252Fcomponents%252FSkillsForm.js%2522%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522spaces%2522%253A%257B%2522clbtf86qj00i53b6jwfbm4yay%2522%253A%257B%2522key%2522%253A%2522clbtf86qj00i53b6jwfbm4yay%2522%252C%2522name%2522%253A%2522Default%2522%252C%2522devtools%2522%253A%255B%257B%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clbtfkgfw00218bhvhrypf0c2%2522%252C%2522key%2522%253A%2522clbtfkdwr028p3b6j6a7krszj%2522%252C%2522isMinimized%2522%253Afalse%257D%252C%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522key%2522%253A%2522clbtf86qj00i63b6jq42ltwy7%2522%252C%2522isMinimized%2522%253Afalse%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522key%2522%253A%2522clbtfgjva01pw3b6jbyyxd00a%2522%252C%2522isMinimized%2522%253Afalse%257D%255D%257D%257D%252C%2522currentSpace%2522%253A%2522clbtf86qj00i53b6jwfbm4yay%2522%252C%2522spacesOrder%2522%253A%255B%2522clbtf86qj00i53b6jwfbm4yay%2522%255D%257D

**Reply by ranran2121** [reply]
what I would like to achieve is:

1. to know how to set default values when using the Form Provider
2. to display a row with empty fields when I do not have info from the DB
3. to prefill the fields with data from the DB when there are such info in the DB
(The data are fetched in the parent component and passed to the child component(s))
I hope I have expressed myself clearly enough

**Reply by leapful** [reply]
@ranran2121 

1. The best practice with "react-hook-form" is to provide default values on `useForm` hook and then pass it to `FormProvider`.
2. Whenever appending new row with empty field, you will need to provide the default value for the new row.
3. You can use `useEffect` to fetch async data then perform `reset` method from `useForm` hook (which is now getting from `FormProvider`) to set new async data when it finished loading.
